Source/Diagnostics/SliceDiagnostic.cpp

```diff
@@ -0,0 +1,490 @@
+            if ( slice_cc_nd_box.lo(idim) - real_box.lo(idim) >= fac ) {            
+                slice_lo[idim] = floor( ( (slice_cc_nd_box.lo(idim) 
+                                 - (real_box.lo(idim) + fac ) ) 
+                                 / dom_geom[0].CellSize(idim)) + fac * 1E-10);
+                slice_lo2[idim] = ceil( ( (slice_cc_nd_box.lo(idim) 
+                                 - (real_box.lo(idim) + fac) ) 
+                                 / dom_geom[0].CellSize(idim)) - fac * 1E-10 );    
+            }            
+            else {            
+                slice_lo[idim] =  round( (slice_cc_nd_box.lo(idim) 
+                                  - (real_box.lo(idim) ) ) 
+                                  / dom_geom[0].CellSize(idim));
+                slice_lo2[idim] = ceil((slice_cc_nd_box.lo(idim) 
+                                  - (real_box.lo(idim) ) ) 
+                                  / dom_geom[0].CellSize(idim) );
+            }
+ 
+            // flag for interpolation -- if reduced dimension location  //
+            //                           does not align with data point //
+            if ( slice_lo[idim] == slice_lo2[idim]) {            
+               if ( slice_cc_nd_box.lo(idim) - real_box.lo(idim) < fac ) {
+                  interp_lo[idim] = 1;
+               }
+            }
+            else {            
+               interp_lo[idim] = 1;
+            }
+
+            // ncells = 1 if dimension is reduced //
+            slice_hi[idim] = slice_lo[idim] + 1;
+        }
+        else
+        {
+            // moving realbox.lo and reabox.hi to nearest coarsenable grid point //
+            int index_lo = floor(((slice_realbox.lo(idim) - (real_box.lo(idim))) 
+                           / dom_geom[0].CellSize(idim) ) +fac * 1E-10);
+            int index_hi = ceil(((slice_realbox.hi(idim) - (real_box.lo(idim))) 
+                           / dom_geom[0].CellSize(idim) ) - fac * 1E-10);
+
+            bool modify_cr = true;
+
+            while ( modify_cr == true) {            
+                int lo_new = index_lo;
+                int hi_new = index_hi;
+                int mod_lo = index_lo % cr_ratio[idim];
+                int mod_hi = index_hi % cr_ratio[idim];
+                modify_cr = false;
+
+                // To ensure that the index.lo is coarsenable //
+                if ( mod_lo > 0) {
+                   lo_new = index_lo - mod_lo;
+                }
+                // To ensure that the index.hi is coarsenable //
+                if ( mod_hi > 0) {
+                   hi_new = index_hi + (cr_ratio[idim] - mod_hi);
+                }
+
+                // If modified index.hi is > baselinebox.hi, reduce cr ratio // 
+                // and provide more points that asked for                    //
+                if ( (hi_new * dom_geom[0].CellSize(idim)) 
+                      > real_box.hi(idim) - real_box.lo(idim) )
```

Comment:
Would you please help me understand why you compare 
    `hi_new * dom_geom[0].CellSize(idim)` with `real_box.hi(idim) - real_box.lo(idim)`
and not 
    `(hi_new-lo_now) * dom_geom[0].CellSize(idim)` with `real_box.hi(idim) - real_box.lo(idim)`
or
    `hi_new * dom_geom[0].CellSize(idim)` with `real_box.hi(idim)`
? Sorry, I think I'm missing something here.

Reply:
hi_new is generated if the hi coordinate provided in the input is not coarsenable. 

It starts with the definition of index_hi which is already normalized. 
hi_new = index_hi if the value is coarsenable. 
If not, hi_new is modified to the next coarsenable index. 
But there is a chance that the value for hi_new generated lies outside the domain. 
So I multiple hi_new * CellSize and see if it lies within the length of the underlying domain using the above snippet. 

I am only checking if hi_new is contained within the extent of the domain. If not, I change the cr_ratio by a factor of 2. For example :  if the coarsening ratio is 8 and it causes hi_new to move outside the domain, I modify the ratio to 4 and ensure that more data points are provided instead of decreasing the extent of the slice. 

Please let me know if this explanation is not clear. I can also meet offline and clarify.